File: android_app/app/src/main/java/com/oceansentinels/app/data/remote/dto/WeatherDto.kt
package com.oceansentinels.app.data.remote.dto

import com.google.gson.annotations.SerializedName

/**
 * WeatherAPI.com response DTOs
 * Docs: https://www.weatherapi.com/docs/
 */
data class WeatherResponse(
    @SerializedName("location") val location: WeatherLocation,
    @SerializedName("current") val current: CurrentWeather,
    @SerializedName("forecast") val forecast: WeatherForecast? = null
)

data class WeatherLocation(
    @SerializedName("name") val name: String,
    @SerializedName("region") val region: String,
    @SerializedName("country") val country: String,
    @SerializedName("lat") val lat: Double,
    @SerializedName("lon") val lon: Double,
    @SerializedName("tz_id") val tzId: String,
    @SerializedName("localtime") val localtime: String
)

data class CurrentWeather(
    @SerializedName("temp_c") val tempC: Double,
    @SerializedName("temp_f") val tempF: Double,
    @SerializedName("is_day") val isDay: Int,
    @SerializedName("condition") val condition: WeatherCondition,
    @SerializedName("wind_mph") val windMph: Double,
    @SerializedName("wind_kph") val windKph: Double,
    @SerializedName("wind_dir") val windDir: String,
    @SerializedName("pressure_mb") val pressureMb: Double,
    @SerializedName("precip_mm") val precipMm: Double,
    @SerializedName("humidity") val humidity: Int,
    @SerializedName("cloud") val cloud: Int,
    @SerializedName("feelslike_c") val feelslikeC: Double,
    @SerializedName("feelslike_f") val feelslikeF: Double,
    @SerializedName("vis_km") val visKm: Double,
    @SerializedName("uv") val uv: Double,
    @SerializedName("gust_kph") val gustKph: Double
)

data class WeatherCondition(
    @SerializedName("text") val text: String,
    @SerializedName("icon") val icon: String,
    @SerializedName("code") val code: Int
)

data class WeatherForecast(
    @SerializedName("forecastday") val forecastDay: List<ForecastDay>
)

data class ForecastDay(
    @SerializedName("date") val date: String,
    @SerializedName("day") val day: DayWeather,
    @SerializedName("astro") val astro: Astro,
    @SerializedName("hour") val hour: List<HourWeather>
)

data class DayWeather(
    @SerializedName("maxtemp_c") val maxtempC: Double,
    @SerializedName("mintemp_c") val mintempC: Double,
    @SerializedName("avgtemp_c") val avgtempC: Double,
    @SerializedName("maxwind_kph") val maxwindKph: Double,
    @SerializedName("totalprecip_mm") val totalprecipMm: Double,
    @SerializedName("avghumidity") val avghumidity: Double,
    @SerializedName("daily_chance_of_rain") val dailyChanceOfRain: Int,
    @SerializedName("condition") val condition: WeatherCondition,
    @SerializedName("uv") val uv: Double
)

data class Astro(
    @SerializedName("sunrise") val sunrise: String,
    @SerializedName("sunset") val sunset: String,
    @SerializedName("moonrise") val moonrise: String,
    @SerializedName("moonset") val moonset: String,
    @SerializedName("moon_phase") val moonPhase: String
)

data class HourWeather(
    @SerializedName("time") val time: String,
    @SerializedName("temp_c") val tempC: Double,
    @SerializedName("condition") val condition: WeatherCondition,
    @SerializedName("wind_kph") val windKph: Double,
    @SerializedName("humidity") val humidity: Int,
    @SerializedName("feelslike_c") val feelslikeC: Double,
    @SerializedName("chance_of_rain") val chanceOfRain: Int,
    @SerializedName("is_day") val isDay: Int
)

data class MarineResponse(
    @SerializedName("location") val location: WeatherLocation,
    @SerializedName("forecast") val forecast: MarineForecast
)

data class MarineForecast(
    @SerializedName("forecastday") val forecastDay: List<MarineForecastDay>
)

data class MarineForecastDay(
    @SerializedName("date") val date: String,
    @SerializedName("day") val day: DayWeather,
    @SerializedName("hour") val hour: List<MarineHourWeather>
)

data class MarineHourWeather(
    @SerializedName("time") val time: String,
    @SerializedName("temp_c") val tempC: Double,
    @SerializedName("condition") val condition: WeatherCondition,
    @SerializedName("wind_kph") val windKph: Double,
    @SerializedName("humidity") val humidity: Int,
    @SerializedName("sig_ht_mt") val sigHtMt: Double? = null,
    @SerializedName("swell_ht_mt") val swellHtMt: Double? = null,
    @SerializedName("swell_dir") val swellDir: String? = null,
    @SerializedName("swell_period_secs") val swellPeriodSecs: Double? = null,
    @SerializedName("water_temp_c") val waterTempC: Double? = null,
    @SerializedName("tide_height_mt") val tideHeightMt: String? = null
)

File: IOS/OceanSentinels/Data/Remote/DTO/WeatherDTO.swift
import Foundation

// MARK: - WeatherAPI.com DTOs

struct WeatherResponse: Codable {
    let location: WeatherLocation
    let current: CurrentWeather
    let forecast: WeatherForecast?
}

struct WeatherLocation: Codable {
    let name: String
    let region: String
    let country: String
    let lat: Double
    let lon: Double
    let tzId: String
    let localtime: String
    
    enum CodingKeys: String, CodingKey {
        case name, region, country, lat, lon
        case tzId = "tz_id"
        case localtime
    }
}

struct CurrentWeather: Codable {
    let tempC: Double
    let tempF: Double
    let isDay: Int
    let condition: WeatherCondition
    let windMph: Double
    let windKph: Double
    let windDir: String
    let pressureMb: Double
    let precipMm: Double
    let humidity: Int
    let cloud: Int
    let feelslikeC: Double
    let feelslikeF: Double
    let visKm: Double
    let uv: Double
    let gustKph: Double
    
    enum CodingKeys: String, CodingKey {
        case tempC = "temp_c"
        case tempF = "temp_f"
        case isDay = "is_day"
        case condition
        case windMph = "wind_mph"
        case windKph = "wind_kph"
        case windDir = "wind_dir"
        case pressureMb = "pressure_mb"
        case precipMm = "precip_mm"
        case humidity, cloud
        case feelslikeC = "feelslike_c"
        case feelslikeF = "feelslike_f"
        case visKm = "vis_km"
        case uv
        case gustKph = "gust_kph"
    }
}

struct WeatherCondition: Codable {
    let text: String
    let icon: String
    let code: Int
}

struct WeatherForecast: Codable {
    let forecastDay: [ForecastDay]
    
    enum CodingKeys: String, CodingKey {
        case forecastDay = "forecastday"
    }
}

struct ForecastDay: Codable, Identifiable {
    var id: String { date }
    let date: String
    let day: DayWeather
    let astro: Astro
    let hour: [HourWeather]
}

struct DayWeather: Codable {
    let maxtempC: Double
    let mintempC: Double
    let avgtempC: Double
    let maxwindKph: Double
    let totalprecipMm: Double
    let avghumidity: Double
    let dailyChanceOfRain: Int
    let condition: WeatherCondition
    let uv: Double
    
    enum CodingKeys: String, CodingKey {
        case maxtempC = "maxtemp_c"
        case mintempC = "mintemp_c"
        case avgtempC = "avgtemp_c"
        case maxwindKph = "maxwind_kph"
        case totalprecipMm = "totalprecip_mm"
        case avghumidity
        case dailyChanceOfRain = "daily_chance_of_rain"
        case condition, uv
    }
}

struct Astro: Codable {
    let sunrise: String
    let sunset: String
    let moonrise: String
    let moonset: String
    let moonPhase: String
    
    enum CodingKeys: String, CodingKey {
        case sunrise, sunset, moonrise, moonset
        case moonPhase = "moon_phase"
    }
}

struct HourWeather: Codable, Identifiable {
    var id: String { time }
    let time: String
    let tempC: Double
    let condition: WeatherCondition
    let windKph: Double
    let humidity: Int
    let feelslikeC: Double
    let chanceOfRain: Int
    let isDay: Int
    
    enum CodingKeys: String, CodingKey {
        case time
        case tempC = "temp_c"
        case condition
        case windKph = "wind_kph"
        case humidity
        case feelslikeC = "feelslike_c"
        case chanceOfRain = "chance_of_rain"
        case isDay = "is_day"
    }
}

// MARK: - Marine DTOs

struct MarineResponse: Codable {
    let location: WeatherLocation
    let forecast: MarineForecast
}

struct MarineForecast: Codable {
    let forecastDay: [MarineForecastDay]
    
    enum CodingKeys: String, CodingKey {
        case forecastDay = "forecastday"
    }
}

struct MarineForecastDay: Codable, Identifiable {
    var id: String { date }
    let date: String
    let day: DayWeather
    let hour: [MarineHourWeather]
}

struct MarineHourWeather: Codable, Identifiable {
    var id: String { time }
    let time: String
    let tempC: Double
    let condition: WeatherCondition
    let windKph: Double
    let humidity: Int
    let sigHtMt: Double?
    let swellHtMt: Double?
    let swellDir: String?
    let swellPeriodSecs: Double?
    let waterTempC: Double?
    let tideHeightMt: String?
    
    enum CodingKeys: String, CodingKey {
        case time
        case tempC = "temp_c"
        case condition
        case windKph = "wind_kph"
        case humidity
        case sigHtMt = "sig_ht_mt"
        case swellHtMt = "swell_ht_mt"
        case swellDir = "swell_dir"
        case swellPeriodSecs = "swell_period_secs"
        case waterTempC = "water_temp_c"
        case tideHeightMt = "tide_height_mt"
    }
}
